### PR TITLE
Test jira

### PR DESCRIPTION
**Pull request by mjagelka**
I added some changes to the jira.py code, which I will explain in comments. I also added unit tests for this file.

### CODE REVIEW

**Comment by mjagelka** [comment]
I actually added all of previous comments right after creating pull request. I just did not submit it, so you could not see that. I will just respond to that in your comments.

File: ticketutil/jira.py [hunk 1]
```diff
@@ -359,23 +355,16 @@ def add_watcher(self, watcher):
             watcher = "{0}".format(watcher.split('@')[0].strip())
         watcher = "\"{0}\"".format(watcher)
 
-        # For some reason, if you try to add an empty string as a watcher, it adds the requestor.
-        # So, only execute this code if the watcher is not an empty string.
-        if watcher:
```
[the remark this answers is not in the document]
Comment (mjagelka) [reply ticketutil/jira.py]: As I mentioned in one comment, this does not work as expected.

File: ticketutil/jira.py [hunk 2]
```diff
@@ -65,9 +62,7 @@ def _verify_project(self, project):
             logging.debug("Project {0} is valid".format(project))
             return True
         except requests.RequestException as e:
-            if r.json()['errorMessages'][0] == "No project could be found with key \'{0}\'.".format(project):
```
[the remark this answers is not in the document]
Comment (mjagelka) [reply ticketutil/jira.py]: As I mentioned in a comment bellow, this is a duplicate.

File: ticketutil/jira.py [hunk 3]
```diff
@@ -85,42 +80,43 @@ def _verify_ticket_id(self, ticket_id):
             logging.debug("Ticket {0} is valid".format(ticket_id))
             return True
         except requests.RequestException as e:
-            if r.json()['errorMessages'][0] == "Issue Does Not Exist":
```
[the remark this answers is not in the document]
Comment (mjagelka) [reply ticketutil/jira.py]: The same problem is here.

File: tests/test_jira.py [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+STATUS = {'transitions': [{'to': {'name': 'To Do'}, 'id': 11}, {'to': {'name': 'Abandoned'}, 'id': 51}]}
+WATCHERS = {'watchers': [{'name': 'me'}, {'name': 'you'}]}
+
+
+class FakeSession(object):
+    """
+    Mocks Requests session behavior.
+    """
+
+    def __init__(self, status_code=666):
```
[the remark this answers is not in the document]
Comment (mjagelka) [reply tests/test_jira.py]: Is it a problem to continue in this trend? Or did you want to point out that it is inappropriate to have a devil's number here? :D

File: ticketutil/jira.py [hunk 5]
```diff
@@ -311,7 +307,7 @@ def remove_all_watchers(self):
             return self.request_result._replace(status='Failure', error_message=error_message)
         else:
             logging.info("Removed watchers from ticket {0} - {1}".format(self.ticket_id, self.ticket_url))
-            return self.request_result._replace(watchers=watchers_list)
+            return self.request_result
```
[the remark this answers is not in the document]
Comment (mjagelka) [reply ticketutil/jira.py]: Ok, I see now how it works. I changed it back.

File: ticketutil/jira.py [hunk 6]
```diff
@@ -85,42 +80,43 @@ def _verify_ticket_id(self, ticket_id):
...
-            if r.json()['errorMessages'][0] == "Issue Does Not Exist":
-                logging.error("Ticket {0} is not valid".format(ticket_id))
-            else:
+            if r.json()['errorMessages'][0] != "Issue Does Not Exist":
                 logging.error("Unexpected error occurred when verifying ticket_id")
                 logging.error(e)
             return False
 
-    def create(self, summary, description, **kwargs):
+    def create(self, summary, description, issue_type, **kwargs):
```
[the remark this answers is not in the document]
Comment (mjagelka) [reply ticketutil/jira.py]: I changed `issue_type` to `type`. I thought about having the argument named `type` from the beginning but it collides with python buildin function, so later i decided for different name.

[the remark this answers is not in the document]
Comment (mjagelka) [reply ticketutil/jira.py]: I tested the necessity of `issue_type` field on stage JIRA instance. I can also try it on production. Are there other instances I am not aware of?

**Comment by dmranck** [comment]
Just wanted to let you know - sorry for the delay with my review... This is on my list to get to in the next couple days.

File: ticketutil/jira.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dmranck) [reply ticketutil/jira.py]: See comment below.

File: ticketutil/jira.py [hunk 7]
```diff
@@ -65,9 +62,7 @@ def _verify_project(self, project):
             logging.debug("Project {0} is valid".format(project))
             return True
         except requests.RequestException as e:
-            if r.json()['errorMessages'][0] == "No project could be found with key \'{0}\'.".format(project):
-                logging.error("Project {0} is not valid".format(project))
-            else:
+            if r.json()['errorMessages'][0] != "No project could be found with key \'{0}\'.".format(project):
```
[the remark this answers is not in the document]
Comment (dmranck) [reply ticketutil/jira.py]: I disagree here. ticket.py only raises the TicketError with the "Project __ is not valid" message. It doesn't **log** an error message. The original code in _verify_project() here checks for and then logs two separate errors:

- _Checks for:_ "No project could be found with key \'{0}\'.". _Logs:_ "Project {0} is not valid"
- _Checks for:_ An unexpected error.  _Logs:_ "Unexpected error occurred when verifying project"

This is giving the user a more specific error message - either a project couldn't be found with the specified key, or something else went wrong. Regardless of which error message is logged, the same TicketException is raised - "Project __ is not valid".

I'd still like to see this part reverted back to the original code.

File: ticketutil/jira.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmranck) [reply ticketutil/jira.py]: This isn't a problem - it's by design. Same comment as with _verify_project() - the original code checks for and logs two separate errors, then raises the TicketException with "Ticket {0} is not valid". 

There's no code duplication in these two methods - it's just a matter of checking for different possible errors that could be returned from the JIRA API, logging those different errors, and then raising a common TicketException. 

Hope that makes sense.

File: ticketutil/jira.py [hunk 8]
```diff
@@ -359,23 +355,16 @@ def add_watcher(self, watcher):
...
-                logging.debug("Add watcher {0}: status code: {1}".format(watcher, r.status_code))
-                r.raise_for_status()
-                logging.info("Added watcher {0} to ticket {1} - {2}".format(watcher, self.ticket_id, self.ticket_url))
-                return self.request_result
-            except requests.RequestException as e:
-                error_message = "Error adding {0} as a watcher to ticket".format(watcher)
-                logging.error(error_message)
-                logging.error(e)
-                return self.request_result._replace(status='Failure', error_message=error_message)
-        else:
```
[the remark this answers is not in the document]
Comment (dmranck) [reply ticketutil/jira.py]: My follow-up question here is what "didn't allow it to pass" when you tried passing an empty string in? The JIRA API or the ticketutil code? When I tested this by passing in an empty string, the requestor was added as a watcher. What exactly happened when you tested it?

File: ticketutil/jira.py [hunk 9]
```diff
@@ -41,9 +41,6 @@ def _generate_ticket_url(self):
 
         # If we are receiving a ticket_id, it indicates we'll be doing an update or resolve, so set ticket_url.
         if self.ticket_id:
-            if '-' not in self.ticket_id:
```
[the remark this answers is not in the document]
Comment (dmranck) [reply ticketutil/jira.py]: Oh yeah, if you rebase onto the latest master, you'll get this update too :)

File: ticketutil/jira.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dmranck) [reply ticketutil/jira.py]: See comment below.

[the remark this answers is not in the document]
Comment (mjagelka) [reply ticketutil/jira.py]: Changes were reverted. Quote wrapping of watchers was moved inside `if` clause.

File: ticketutil/jira.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mjagelka) [reply ticketutil/jira.py]: Changes were reverted.

File: ticketutil/jira.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mjagelka) [reply ticketutil/jira.py]: Changes were reverted.

File: ticketutil/jira.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mjagelka) [reply ticketutil/jira.py]: My mistake. When I tested empty watcher, it raised the error (I don't recall which kind of error), therefore I thought it doesn't allow to pass empty string. However when I tested that once again, it worked like you described. I reverted this change and adjusted the `if` clause.

File: ticketutil/jira.py [hunk 10]
```diff
@@ -80,47 +77,50 @@ def _verify_ticket_id(self, ticket_id):
...
-            logging.debug("Verify ticket_id: status code: {0}".format(r.status_code))
+            logger.debug("Verify ticket_id: status code: {0}".format(r.status_code))
             r.raise_for_status()
-            logging.debug("Ticket {0} is valid".format(ticket_id))
+            logger.debug("Ticket {0} is valid".format(ticket_id))
             return True
         except requests.RequestException as e:
             if r.json()['errorMessages'][0] == "Issue Does Not Exist":
-                logging.error("Ticket {0} is not valid".format(ticket_id))
+                logger.error("Project {0} is not valid".format(project))
```
[the remark this answers is not in the document]
Comment (mjagelka) [reply ticketutil/jira.py]: My fault. It's changed now.

File: ticketutil/jira.py [hunk 11]
```diff
@@ -91,35 +91,38 @@ def _verify_ticket_id(self, ticket_id):
...
+        :param type: The ticket issue type.
         :return: self.request_result: Named tuple containing request status, error_message, and url info.
         """
         error_message = ""
         if summary is None:
             error_message = "summary is a necessary parameter for ticket creation"
         if description is None:
             error_message = "description is a necessary parameter for ticket creation"
+        if type is None:
+            error_message = "issue type is a necessary parameter for ticket creation"
```
Comment (dmranck) [inline ticketutil/jira.py]: Let's change 'issue type' here to simply 'type', to not confuse the user into thinking they need to pass in a parameter called 'issue type'.

File: ticketutil/jira.py [hunk 12]
```diff
@@ -91,35 +91,38 @@ def _verify_ticket_id(self, ticket_id):
                 logger.error(e)
             return False
 
-    def create(self, summary, description, **kwargs):
+    def create(self, summary, description, type, **kwargs):
         """
         Creates a ticket.
-        The required parameters for ticket creation are summary and description.
+        The required parameters for ticket creation are summary, description and issue type.
```
Comment (dmranck) [inline ticketutil/jira.py]: Change 'issue type' to 'type'.

File: ticketutil/jira.py [hunk 13]
```diff
@@ -91,35 +91,38 @@ def _verify_ticket_id(self, ticket_id):
...
         # Create our ticket.
         return self._create_ticket_request(params)
 
-    def _create_ticket_parameters(self, summary, description, fields):
+    def _create_ticket_parameters(self, summary, description, type, fields):
         """
         Creates the payload for the POST request when creating a JIRA ticket.
 
-        The required parameters for ticket creation are summary and description.
+        The required parameters for ticket creation are summary, description and issue type.
```
Comment (dmranck) [inline ticketutil/jira.py]: Change 'issue type' to 'type'.

File: read-the-docs/source/JIRA.rst [hunk 14]
```diff
@@ -28,7 +28,7 @@ Methods
 create()
 --------
 
-``create(self, summary, description, **kwargs)``
+``create(self, summary, description, type, **kwargs)``
 
 Creates a ticket. The required parameters for ticket creation are
 summary and description. Keyword arguments are used for other ticket
```
[the remark this answers is not in the document]
Comment (mjagelka) [reply read-the-docs/source/JIRA.rst]: Corrected. Sorry for this inattention.

File: ticketutil/jira.py (hunk 12 above)
Comment (mjagelka) [reply ticketutil/jira.py]: Changed.

File: ticketutil/jira.py (hunk 13 above)
Comment (mjagelka) [reply ticketutil/jira.py]: Changed.

**Comment by mjagelka** [comment]
I am ready.

**Comment by dmranck** [comment]
Thanks, @mjagelka!

**Pull request merged**